### PR TITLE
Implement timeout system

### PR DESCRIPTION
**Pull request by connor4312**
This adds the backend implementation for the timeout system.

Note that merge must wait until the `parse-duration` dependency is approved.

### CODE REVIEW

File: timeout/chat/timeout.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+     * Notifies the user that they're timed out until the provided date,
+     * and updates the associated channel locally.
+     * @param  {User} user
+     * @param  {Date|String} until
+     */
+    _setTimeout(user, until) {
+        until = until instanceof Date ? until : new Date(until);
+
+        user.socket.sendEvent(event, {
+            user: user.getId(),
```
Comment (artdude543) [inline timeout/chat/timeout.js]: Could we have the username also sent in the packet? (Would fit with the rest of the other packets which send user data) Assuming this is the packet which is being sent on the Socket.

Comment (connor4312) [inline timeout/chat/timeout.js]: Sure, done 😉

**Comment by Antariano** [comment]
lgtm

File: timeout/chat/timeout.js [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        if (!channel.chatTimeouts) {
+            channel.chatTimeouts = {};
+        }
+
+        channel.chatTimeouts[user.getId()] = until;
+    }
+
+    /**
+     * onChannel is invoked when a new channel is created on the chat server.
+     * @param  {[type]} channel [description]
```
Comment (SimonSchick) [inline timeout/chat/timeout.js]: doc placeholders.

File: timeout/chat/timeout.js [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+/**
+ * Timeout handles management of user timeouts for channels on the chat server.
+ * While users are timed out, they're unable to chat.
+ */
+class Timeout {
+
+    /**
+     * Creates a new Timeout instance.
+     * @param {MySLQL} sql
```
Comment (SimonSchick) [inline timeout/chat/timeout.js]: typo

File: test/fixture/user.js [hunk 4]
```diff
@@ -8,7 +9,15 @@ module.exports = function (channel) {
...
     user.permissions = [];
+    user.toObject = () => ({
+        user_id: 42,
+        user_name: 'connor4312',
+        user_roles: ['Developer', 'User'],
+    });
     user.hasPermission = function (p) { return this.permissions.indexOf(p) !== -1; };
+    user.socket = socket();
+
+    channel.users = { 'u42': [user] };
```
Comment (SimonSchick) [inline test/fixture/user.js]: Doesn't need quotes.

**Comment by SimonSchick** [comment]
Minor things, looks fine.

**Pull request merged**